File: javascript/word-count/word-count.js
export class Words {
  count(phrase) {
    return phrase
      .trim()
      .replace(/\s+/g, " ")
      .split(" ")
      .map(word => word.toString().toLowerCase())
      .reduce(
        (acc, curr) =>
          acc.hasOwnProperty(curr)
            ? Object.assign(acc, { [curr]: acc[curr] + 1 })
            : Object.assign(acc, { [curr]: 1 }),
        {}
      );
  }
}

File: ruby/word-count/word_count.rb
class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    @phrase
      .split(' ')
      .map { |word| downcase_and_remove_trailing_symbols(word) }
      .each_with_object({}) do |word, acc|
        acc[word] ? acc[word] += 1 : acc[word] = 1
      end
  end

  private

  def downcase_and_remove_trailing_symbols(word)
    word.downcase.gsub(/(^[\W+]|\W*$)/, '')
  end
end
